### PR TITLE
Helios sp1 integration

### PR DESCRIPTION
**Pull request by 0xSasaPrsic**
# Pull Request type

Helios sp1 integration.
https://github.com/succinctlabs/sp1-helios/blob/main/contracts/src/SP1LightClient.sol

Please add the labels corresponding to the type of changes your PR introduces:

- [x] Feature
- [ ] Bugfix
- [ ] Refactor
- [ ] Format
- [ ] Documentation
- [ ] Testing
- [ ] Other:

## Checklist
- [ ] I have performed a self-review of my own code.
- [ ] The tests pass successfully with `cargo test`.
- [ ] The code was formatted with `cargo fmt`.
- [ ] The code compiles with no new warnings with `cargo build --release` and `cargo build --release --features runtime-benchmarks`.
- [ ] The code has no new warnings when using `cargo clippy`.
- [ ] If this change affects documented features or needs new documentation, I have created a PR with a [documentation update](https://github.com/availproject/availproject.github.io).

### CODE REVIEW

File: pallets/vector/Cargo.toml [hunk 1]
```diff
@@ -20,6 +20,9 @@ sp-runtime = { workspace = true, default-features = false }
 sp-std = { workspace = true, default-features = false }
 sp-core = { workspace = true, default-features = false }
 frame-benchmarking = { workspace = true, default-features = false, optional = true }
+# TODO is there a release version?
```
Comment (ratankaliani) [inline pallets/vector/Cargo.toml:23]: The `4.0.0-rc1` tag should be out in the next few days, which you can use. For now, I'd recommend pinning to a specific commit.

File: pallets/vector/Cargo.toml [hunk 2]
```diff
@@ -38,6 +41,7 @@ ethabi.workspace = true
 [dev-dependencies]
 pallet-balances = { workspace = true, default-features = false, features = ["std"] }
 pallet-timestamp = { workspace = true, default-features = false }
+sp1-sdk = { git = "https://github.com/succinctlabs/sp1", branch = "dev", default-features = false}
```
Comment (ratankaliani) [inline pallets/vector/Cargo.toml:44]: Same here.

File: pallets/vector/src/benchmarking.rs [hunk 3]
```diff
@@ -164,13 +164,16 @@ pub const ROTATE_VK: &str = r#"{"vk_json":{
     ]
 }}"#;
 
+const SP1_VERIFICATION_KEY: [u8; 32] =
```
Comment (ratankaliani) [inline pallets/vector/src/benchmarking.rs:167]: Add a description for how this is generated.

File: pallets/vector/src/lib.rs [hunk 4]
```diff
@@ -780,6 +814,127 @@ pub mod pallet {
...
+
+			let is_valid = Groth16Verifier::verify(
+				&proof,
+				&public_values,
+				&format!("{:?}", sp1_vk),
+				&GROTH16_VK_BYTES,
+			);
+			ensure!(is_valid.is_ok(), Error::<T>::VerificationFailed);
+
+			Head::<T>::set(new_head);
```
Comment (ratankaliani) [inline pallets/vector/src/lib.rs:850]: This seems correct.

File: pallets/vector/src/mock.rs [hunk 5]
```diff
@@ -232,9 +232,9 @@ pub fn new_test_ext() -> sp_io::TestExternalities {
 	.unwrap();
 
 	vector_bridge::GenesisConfig::<Test> {
-		finality_threshold: 461,
+		// finality_threshold: 461,
```
Comment (ratankaliani) [inline pallets/vector/src/mock.rs:235]: nit: These should be deleted, rather than commented.

File: pallets/vector/src/tests.rs [hunk 6]
```diff
@@ -1372,3 +1377,355 @@ fn update_updater_non_root() {
...
+			BoundedVec::truncate_from(proof),
+			BoundedVec::truncate_from(public_inputs),
+		);
+
+		// period is zero and sync committee is not set
+		assert_err!(err, Error::<Test>::ConfigurationNotSet);
+	});
+}
+
+// #[test]
```
Comment (ratankaliani) [inline pallets/vector/src/tests.rs:1671]: delete empty test

**Review by Leouarz — APPROVED** [review]
LGTM with overall looking since i don't have all the context.
2 libs still need real version instead of revisions.
Are benchmarks done in the reference machine ?

**Comment by 0xSasaPrsic** [comment]
> LGTM with overall looking since i don't have all the context. 2 libs still need real version instead of revisions. Are benchmarks done in the reference machine ?

Correct, it needs running benchmarks on the referent machine and re-running build API once the review changes are done on the pallet.

**Review by ToufeeqP — APPROVED** [review]

File: pallets/vector/Cargo.toml [hunk 7]
```diff
@@ -20,6 +20,9 @@ sp-runtime = { workspace = true, default-features = false }
 sp-std = { workspace = true, default-features = false }
 sp-core = { workspace = true, default-features = false }
 frame-benchmarking = { workspace = true, default-features = false, optional = true }
+# TODO add a release version
+sp1-verifier = { git = "https://github.com/succinctlabs/sp1", tag = "v3.3.0", default-features = false }
```
Comment (ratankaliani) [inline pallets/vector/Cargo.toml:24]: Use crates.io for this

version = 3.4.0

File: pallets/vector/Cargo.toml [hunk 8]
```diff
@@ -38,6 +41,8 @@ ethabi.workspace = true
 [dev-dependencies]
 pallet-balances = { workspace = true, default-features = false, features = ["std"] }
 pallet-timestamp = { workspace = true, default-features = false }
+# TODO add a release version
+sp1-sdk = { git = "https://github.com/succinctlabs/sp1",  tag = "v3.3.0", default-features = false }
```
Comment (ratankaliani) [inline pallets/vector/Cargo.toml:45]: Use crates.io for this

File: pallets/vector/src/benchmarking.rs [hunk 9]
```diff
@@ -164,13 +164,18 @@ pub const ROTATE_VK: &str = r#"{"vk_json":{
     ]
 }}"#;
 
+// Generated with SP1 Helios https://github.com/succinctlabs/sp1-helios/blob/main/README.md
+// cargo prove key —-elf (sp1 helios elf) in SP1 Helios
+const SP1_VERIFICATION_KEY: [u8; 32] =
+	hex!("00788ce8dc2970920a3d3c072c8c07843d15f1307a53b3dd31b113c3e71c28e8");
```
Comment (ratankaliani) [inline pallets/vector/src/benchmarking.rs:170]: Note: On the latest `main` for SP1 Helios, we've created reproducible builds with Docker. The latest verification key is `0x1b9571a44d20ab19380ef04e5e01052f3ca90a9a7394ad7b2c8b7a5f4705067b`.

You can reproduce with `cargo run --bin vkey --release` on the latest `sp1-helios` main.

File: pallets/vector/src/lib.rs [hunk 10]
```diff
@@ -780,6 +816,130 @@ pub mod pallet {
...
+		#[pallet::call_index(13)]
+		#[pallet::weight(T::WeightInfo::fulfill())]
+		pub fn fulfill(
+			origin: OriginFor<T>,
+			proof: ProofInput,
+			public_values: PublicValuesInput,
+		) -> DispatchResultWithPostInfo {
+			let sender: [u8; 32] = ensure_signed(origin)?.into();
+			let updater = Updater::<T>::get();
+			// ensure sender is preconfigured
```
Comment (ratankaliani) [inline pallets/vector/src/lib.rs:832]: Nice, permissioned proposer.

File: pallets/vector/src/lib.rs [hunk 11]
```diff
@@ -780,6 +816,130 @@ pub mod pallet {
...
+
+			let sp1_vk = SP1VerificationKey::<T>::get();
+
+			let is_valid = Groth16Verifier::verify(
+				&proof,
+				&public_values,
+				&format!("{:?}", sp1_vk),
+				&GROTH16_VK_BYTES,
+			);
+			ensure!(is_valid.is_ok(), Error::<T>::VerificationFailed);
```
Comment (ratankaliani) [inline pallets/vector/src/lib.rs:850]: Note: Would add a comment here about how this leads to unexpected behavior due to the `unwrap()`'s in v3, but will be resolved in v4.

File: pallets/vector/src/lib.rs
Comment (ratankaliani) [inline pallets/vector/src/lib.rs:1]: LGTM

File: pallets/vector/src/tests.rs [hunk 12]
```diff
@@ -21,13 +21,21 @@ use frame_support::{
...
 use sp_runtime::{testing::H256, traits::BadOrigin};
 
 const TEST_SENDER_VEC: [u8; 32] =
 	hex!("d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d");
 const TEST_SENDER_ACCOUNT: AccountId32 = AccountId32::new(TEST_SENDER_VEC);
 
+// Generated with SP1 Helios https://github.com/succinctlabs/sp1-helios/blob/main/README.md
+// cargo prove key —-elf (sp1 helios elf) in SP1 Helios
+const SP1_VERIFICATION_KEY: [u8; 32] =
+	hex!("00788ce8dc2970920a3d3c072c8c07843d15f1307a53b3dd31b113c3e71c28e8");
```
Comment (ratankaliani) [inline pallets/vector/src/tests.rs:35]: Note: I would consolidate the vkey into the `lib.rs` file so that it's not replicated in several places.

**Review by ratankaliani — APPROVED** [review]
LGTM - left a few small comments.

File: pallets/vector/src/benchmarking.rs (hunk 9 above)
Comment (0xSasaPrsic) [reply pallets/vector/src/benchmarking.rs:170]: This is just for the benchmarks, the vk will be added after the release.

**Review by luka-ethernal — APPROVED** [review]

**Pull request merged**